source/core/Variants.cpp

```diff
@@ -172,25 +173,43 @@ VIREO_FUNCTION_SIGNATURET(VariantToData, VariantToDataParamBlock)
                 if (errPtr)
                     errPtr->SetErrorAndAppendCallChain(true, kVariantIncompatibleType, "Variant To Data");
             } else {
+                DualTypeVisitor visitor;
+                DualTypeConversion dualTypeConversion;
+                bool typesCompatible = false;
                 TypeRef underlyingType = variant->_underlyingTypeRef;
                 if (targetType->IsVariant()) {
-                    if (outputData)
+                    if (outputData) {
                         *static_cast<VariantTypeRef *>(outputData) = VariantType::CreateNewVariantFromType(underlyingType);
-                } else if (underlyingType->IsA(targetType, true)) {
-                    if (outputData)
-                        targetType->CopyData(underlyingType->Begin(kPARead), outputData);
```

Comment:
I tend to think we should delegate to type's CopyData if the types are exact, because the type's CopyData can do an efficient Copy when it can. Either make the Visitor to find out whether they are exact or use IsA first to see whether the types are exact and use type's CopyData. Use visitor to copy only when the types are not exactly same but convertible and thus efficient copies cannot be made. We could make the decision in the Visitor's Apply methods also.

Reply:
I was wrong about `IsA` when I mentioned that it is a strict check. It does treat numeric types as compatible. See [here](https://github.com/ni/VireoSDK/blob/defac12ea6b1634cba2b8cd16d77aac4b4c1bb44/source/core/TypeAndDataManager.cpp#L795). So it isn't really an exact type check. The version of `IsA` without the boolean is even less strict because it doesn't dig through nested types.

Is there a different method that can tell us if it is an exact match? Sounds to me like we should implement a third `operation` for the visitor (like `equals` and `conversion`) that does strict type checking..

Reply:
Created this tech debt item to track this performance optimization - https://ni.visualstudio.com/DevCentral/_workitems/edit/142216

Reply:
> It does treat numeric types as compatible. See here. So it isn't really an exact type check. 

I think it is a strict enough check. It checks for different numeric encoding but it does look for the same size (TopAQSize) for the two types. So, for conversion operations, doing bitwise memory copy for integers of same size whether signed or unsigned should still be correct. I think we can fix the code to do what it used to do when the types are exact as determined by IsA with compatibleStructure as true. 

Basically, my point is that G code trying to use exact types of arrays and clusters of numerics (likely a common use case, especially for code written to use GTypes) should be allowed to do memcpy of the whole data (as the existing code is already doing before this change). Almost exact compatible types but  (as determined by IsA with compatibleStructure as true) seems to identify this already correctly.

The scripting tests we need to write for all permutations of different sizes of signed/unsigned numbers and arrays/clusters of numbers can be used to validate this. So, I am not too worried about making this change.

Reply:
That does sounds correct. I will make this change. 

I am working on the scripting tests for edge case numeric data types now and having this vireo version integrated will make things easier. If I do discover that something is broken, I will come back and fix it here.